### ISSUE TITLE
Can't add columns to live "devices" table in production, because MySQL

### ISSUE THREAD

**Issue by rfk** [mozilla/fxa-auth-db-mysql#273]
This is a fun one.  See https://bugzilla.mozilla.org/show_bug.cgi?id=1392776#c15 for the gory details, but the summary seems to be:

* The `devices` table is unusual, in that it has a unique constraint on `(uid, sessionTokenId)` and our normal traffic patterns produce queries that fail out on this constraint at a small but non-zero background rate.
* MySQL has an unfortunate ~~bug~~ ~~feature~~ behaviour where, if an `ALTER TABLE` is running at the same time as a query that fails out on a unique constraint, the `ALTER TABLE` will also fail and rollback.
* The live production `devices` table is very large, so attempts to alter it take tens of minutes or more to complete.
* Therefore, any attempt to `ALTER TABLE devices` on the live production db has a high chance of failing out due to a concurrent unique constraint violation, causing an expensive rollback and a production outage.

This derailed the deployment of [1] and I would expect it to also derail the deployment of [2].

So we can we do about it?  I see a couple of possibilities in the short term, and would be delighted to hear other suggestions.

1.  We could forbid writes to the `devices` table while applying the migration, e.g. by 503-ing POSTs to the device routes, or by running it with `LOCK=shared`.  This would cause a minor and hopefully-mostly-invisible-to-users service outage, but it's pretty ad-hoc and we have to remember to do it every time.

2. We could give up on adding new columns to the `devices` table, and instead create a new `extraDeviceData` table into which to add the new columns.  That's pretty yuck but we have some precedent for it in the way we added tokenVerificationIds to the `sessionTokens` table.

3. We could back out the two migrations to the `devices` table and revisit them later, e.g. as part of migrating to a whole new database instance.  IIUC we don't have a clear idea of the what or when of such a migration though.

We should also have a broader conversation about the future of MySQL and whether we want to migrate to e.g. Aurora or some other DB longer term.  But we need to do something short-term because we can't deploy the head of this repo right now 😢.

@vladikoff @jrgm @jbuck @mozilla/fxa-devs thoughts?

[1] https://github.com/mozilla/fxa-auth-db-mysql/pull/267
[2] https://github.com/mozilla/fxa-auth-db-mysql/pull/272

**Comment by vbudhram** [comment]
Excellent write up @rfk! My thoughts

* The `tokenVerficationId` table has worked pretty well since we have added it. However, it had a very specific purpose in enabling sign-in confirmation. Making an `extraDeviceData` table would work but I think we might fall into a similar situation if/when that table gets big and we need to add more items.

* I like the idea of using Aurora. Seem's like it could reduce/solve a lot of the problems.

* Would using a tool like [gh-ost](https://github.com/github/gh-ost) or [pt-online-schema-change](https://www.percona.com/doc/percona-toolkit/LATEST/pt-online-schema-change.html) make sense in this situation? Both claim to allow ALTER table without locking them.

**Comment by jbuck** [comment]
Well Aurora (once you enable fast DDL via setting `aurora_lab_mode=1` [docs](http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Aurora.DatabaseEngineUpdates.html#Aurora.DatabaseEngineUpdates.LabMode) is fast:

```
MySQL [fxa]> ALTER TABLE devices ADD COLUMN nameUtf8 VARCHAR(255)
  CHARACTER SET utf8mb4 COLLATE utf8mb4_bin;
Query OK, 0 rows affected (0.17 sec)
Records: 0  Duplicates: 0  Warnings: 0
```

**Comment by jbuck** [comment]
I ran into trouble with gh-ost and I filed a bug: https://github.com/github/gh-ost/issues/484

**Comment by rfk** [comment]
> Well Aurora (once you enable fast DDL via setting aurora_lab_mode=1 docs is fast:
> [...]
> Query OK, 0 rows affected (0.17 sec)

Prize for "understatement of the week" goes to @jbuck, that's blazing!

**Comment by rfk** [comment]
@jbuck what do you think about trying the STR in https://bugzilla.mozilla.org/show_bug.cgi?id=1392776#c15 to see if Aurora exhibits the same behaviour as MySQL in this regard?  IIUC it's a fairly different implementation under the hood so it might not have this same limitation.

**Comment by rfk** [comment]
> We could give up on adding new columns to the devices table,
> and instead create a new extraDeviceData table into which to add the new columns. 

After sleeping on it, I'm leaning towards this as the best short-term option.  It's ugly, but it's almost certain to work and unblock us in the short term, and we can undo the ugliness with a more considered migration to Aurora + gh-ost or some other fresh approach in Q4.

If we were confident gh-ost would "just work" then that would be ideal, as we could just apply the migrations as-is.  But sounds like there might be a few hiccups and a bit of uncertainty around this?

**Comment by vladikoff** [comment]
Seems like there is a progress in https://github.com/github/gh-ost/issues/484 

I am still a bit sold on: 

> We could forbid writes to the devices table while applying the migration, e.g. by 503-ing POSTs to the device routes, or by running it with LOCK=shared

🗡

**Comment by rfk** [comment]
> I am still a bit sold on:
> > We could forbid writes to the devices table while applying the migration

I'm not opposed to eating a partial outage here as a one-off thing, if we're confident that the outage would be limited to attempts to write into the devices table.  But if we go this route, let's make sure we have a plan for how it's not going to have to happen again.

My high-level position is: let's do something with a very high probability of success so that we can unblock train-95.

> or by running it with LOCK=shared

FWIW, I did try using LOCK=shared when reproducing this bug and it seemed to successfully avoid the issue.  So the probability of success here seems high.

**Comment by rfk** [comment]
> I'm not opposed to eating a partial outage here as a one-off thing,
> if we're confident that the outage would be limited to attempts to write into the devices table

Looking at the code, the only things that write to the devices table are the "createDevice" and "updateDevice" database procedures.  The auth-server calls one of these at the following points:

1. When a client does an explicit device registration or update
2. When we create a placeholder device record during /certificate/sign
3. When we detect that a push endpoint has expired

Running the migration with `LOCKED=shared` would cause all of these to block (and likely timeout) while the migration is running.

**Comment by jbuck** [comment]
```
Copy: 51064035/51064035 100.0%; Applied: 159823; Backlog: 5/1000; Time: 11h46m37s(total), 2h13m57s(copy); streamer: mysql-bin-changelog.000
336:2039688; State: postponing cut-over; ETA: due
2017-09-07 16:06:36 INFO Grabbing voluntary lock: gh-ost.175165.lock
2017-09-07 16:06:36 INFO Setting LOCK timeout as 6 seconds
2017-09-07 16:06:36 INFO Looking for magic cut-over table
2017-09-07 16:06:36 INFO Creating magic cut-over table `fxa`.`_devices_del`
2017-09-07 16:06:36 INFO Magic cut-over table created
2017-09-07 16:06:36 INFO Locking `fxa`.`devices`, `fxa`.`_devices_del`
2017-09-07 16:06:36 INFO Tables locked
2017-09-07 16:06:36 INFO Session locking original & magic tables is 175165
2017-09-07 16:06:36 INFO Writing changelog state: AllEventsUpToLockProcessed:1504800396591187945
2017-09-07 16:06:36 INFO Waiting for events up to lock
2017-09-07 16:06:36 INFO Intercepted changelog state AllEventsUpToLockProcessed
2017-09-07 16:06:36 INFO Handled changelog state AllEventsUpToLockProcessed
2017-09-07 16:06:37 INFO Waiting for events up to lock: got AllEventsUpToLockProcessed:1504800396591187945
2017-09-07 16:06:37 INFO Done waiting for events up to lock; duration=587.498067ms
# Migrating `fxa`.`devices`; Ghost table is `fxa`.`_devices_gho`
# Migrating ip-10-3-2-230:3306; inspecting ip-10-3-2-65:3306; executing on ip-172-31-11-80
# Migration started at Thu Sep 07 04:19:58 +0000 2017
# chunk-size: 1000; max-lag-millis: 15000ms; dml-batch-size: 10; max-load: Threads_running=25; critical-load: Threads_running=1000; nice-ratio: 0.000000
# throttle-additional-flag-file: /tmp/gh-ost.throttle
# postpone-cut-over-flag-file: /tmp/ghost.postpone.flag
# panic-flag-file: /tmp/ghost.panic.flag
# Serving on unix socket: /tmp/gh-ost.fxa.devices.sock
Copy: 51064035/51064035 100.0%; Applied: 159836; Backlog: 0/1000; Time: 11h46m39s(total), 2h13m57s(copy); streamer: mysql-bin-changelog.000336:2076498; State: migrating; ETA: due
2017-09-07 16:06:37 INFO Setting RENAME timeout as 3 seconds
2017-09-07 16:06:37 INFO Session renaming tables is 175168
2017-09-07 16:06:37 INFO Issuing and expecting this to block: rename /* gh-ost */ table `fxa`.`devices` to `fxa`.`_devices_del`, `fxa`.`_devices_gho` to `fxa`.`devices`
2017-09-07 16:06:37 INFO Found atomic RENAME to be blocking, as expected. Double checking the lock is still in place (though I don't strictly have to)
2017-09-07 16:06:37 INFO Checking session lock: gh-ost.175165.lock
2017-09-07 16:06:37 INFO Connection holding lock on original table still exists
2017-09-07 16:06:37 INFO Will now proceed to drop magic table and unlock tables
2017-09-07 16:06:37 INFO Dropping magic cut-over table
2017-09-07 16:06:37 INFO Releasing lock from `fxa`.`devices`, `fxa`.`_devices_del`
2017-09-07 16:06:37 INFO Tables unlocked
2017-09-07 16:06:37 INFO Tables renamed
2017-09-07 16:06:37 INFO Lock & rename duration: 808.302437ms. During this time, queries on `devices` were blocked
2017-09-07 16:06:37 INFO Looking for magic cut-over table
2017-09-07 16:06:37 INFO Closed streamer connection. err=<nil>
2017-09-07 16:06:37 INFO Droppping table `fxa`.`_devices_ghc`
2017-09-07 16:06:37 INFO Table dropped
2017-09-07 16:06:37 INFO Am not dropping old table because I want this operation to be as live as possible. If you insist I should do it, please add `--ok-to-drop-table` next time. But I prefer you do not. To drop the old table, issue:
2017-09-07 16:06:37 INFO -- drop table `fxa`.`_devices_del`
2017-09-07 16:06:37 INFO Done migrating `fxa`.`devices`
2017-09-07 16:06:37 INFO Removing socket file: /tmp/gh-ost.fxa.devices.sock
# Done
```

🎉 🎉 🎉

**Comment by jbuck** [comment]
The options I used with gh-ost were:

```
gh-ost \
--max-load=Threads_running=25 \
--assume-rbr \
--critical-load=Threads_running=1000 \
--chunk-size=1000 \
--max-lag-millis=15000 \
--user="user" \
--password="password" \
--host="read replica host" \
--assume-master-host="master host" \
--database="fxa" \
--table="devices" \
--verbose \
--alter="ADD COLUMN nameUtf8 VARCHAR(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin" \
--cut-over=default \
--exact-rowcount \
--concurrent-rowcount \
--default-retries=120 \
--panic-flag-file=/tmp/ghost.panic.flag \
--postpone-cut-over-flag-file=/tmp/ghost.postpone.flag \
--execute
```

**Comment by vladikoff** [comment]
Yay!
